h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/predict/XGBoostNativeBigScoreChunkPredict.java

```diff
@@ -0,0 +1,128 @@
+  private final BoosterParms _boosterParms;
+  private final XGBoostOutput _output;
+
+  private final float[][] _preds;
+
+  public XGBoostNativeBigScoreChunkPredict(
+      XGBoostModelInfo modelInfo,
+      XGBoostModel.XGBoostParameters parms,
+      DataInfo di,
+      BoosterParms boosterParms,
+      double threshold,
+      XGBoostOutput output,
+      Frame fr,
+      Chunk[] chks
+  ) {
+    _modelInfo = modelInfo;
+    _parms = parms;
+    _dataInfo = di;
+    _boosterParms = boosterParms;
+    _threshold = threshold;
+    _output = output;
+    
+    _responseIndex = fr.find(_parms._response_column);
+    _offsetIndex = fr.find(_parms._offset_column);
+    _preds = scoreChunk(chks, XGBoostPredict.OutputType.PREDICT);
+  }
+
+  @Override
+  public double[] score0(Chunk[] chks, double offset, int row_in_chunk, double[] tmp, double[] preds) {
+    for (int i = 0; i < tmp.length; i++) {
+      tmp[i] = chks[i].atd(row_in_chunk);
+    }
+    return XGBoostMojoModel.toPreds(tmp, _preds[row_in_chunk], preds, _output.nclasses(), null, _threshold);
+  }
+
+  @Override
+  public float[][] predictContrib(Chunk[] cs) {
+    return scoreChunk(cs, OutputType.PREDICT_CONTRIB_APPROX);
+  }
+
+  @Override
+  public float[][] predict(Chunk[] cs) {
+    return scoreChunk(cs, OutputType.PREDICT);
+  }
+
+  private float[][] scoreChunk(final Chunk[] cs, final XGBoostPredict.OutputType outputType) {
+    DMatrix data = null;
+    Booster booster = null;
+    try {
+      // Rabit has to be initialized as parts of booster.predict() are using Rabit
+      // This might be fixed in future versions of XGBoost
+      Rabit.init(new HashMap<>());
+      data = XGBoostUtils.convertChunksToDMatrix(
+          _dataInfo,
+          cs,
+          _responseIndex,
+          _output._sparse,
+          _offsetIndex
+      );
+      // No local chunks for this frame
+      if (data.rowNum() == 0) {
+        return null;
+      }
+      // Initialize Booster
+      booster = _modelInfo.deserializeBooster();
+      booster.setParams(_boosterParms.get());
+      int treeLimit = 0;
+      if (_parms._booster == XGBoostModel.XGBoostParameters.Booster.dart) {
+        // DART with treeLimit=0 returns non-deterministic random predictions
+        treeLimit = _parms._ntrees;
+      }
+
+      // Predict
+      float[][] preds;
+      switch (outputType) {
+        case PREDICT:
+          preds = booster.predict(data, false, treeLimit);
+          break;
+        case PREDICT_CONTRIB_APPROX:
+          preds = booster.predictContrib(data, treeLimit);
+          break;
+        default:
+          throw new UnsupportedOperationException("Unsupported output type: " + outputType);
+      }
+      return preds == null ? new float[0][] : preds;
+    } catch (XGBoostError xgBoostError) {
+      throw new IllegalStateException("Failed to score with XGBoost.", xgBoostError);
+    } finally {
+      BoosterHelper.dispose(booster, data);
+      try {
+        Rabit.shutdown();
```

Comment:
Sounds like a too much of a paranoia, but what if BoosterHelper fails ? The rabit shutdown could still be a handful thing to do - therefore I'd put it into a finally block as well. Inception really, I know.

Reply:
if the disposing of a booster fails something is very wrong and the whole cluster is probably unhealthy, I dont think we need to worry about shutting down rabit